tests/integration_tests/test_upgrade.py

```diff
@@ -0,0 +1,43 @@
+from tests.integration_tests.clouds import IntegrationCloud
+from tests.integration_tests.conftest import get_validated_source
+import pytest
+
+
+def _output_to_compare(instance):
+    commands = [
+        'hostname',
+        'dpkg-query --show cloud-init',
+        'cat /run/cloud-init/result.json',
+        '! grep Trace /var/log/cloud-init.log',
+        'systemd-analyze',
+        'systemd-analyze blame',
+        'cloud-init analyze show',
+        'cloud-init analyze blame',
+        'cat $NETCFG_FILE',
+        'cloud-id'
+    ]
+    for command in commands:
+        print('executing: {}'.format(command))
+        print(instance.execute(command))
+
+
+def test_upgrade(session_cloud: IntegrationCloud):
+    try:
+        install_method = get_validated_source()
+    except ValueError:
```

Comment:
This is, of course, yet to be updated, but I wonder if we want to encapsulate the check we're performing here in `CloudInitSource`. Something like `if not install_method.performs_full_install(): pytest.skip(...)`?

(I'm wary of maintaining lists of these around the codebase, because it makes it very easy to fail to update things as we introduce new sources; if the list is maintained in `CloudInitSource` then it's harder to miss.)